appserver/concurrent/concurrent-impl/src/main/java/org/glassfish/concurrent/runtime/ContextSetupProviderImpl.java

```diff
@@ -149,11 +147,21 @@ public ContextHandle setup(ContextHandle contextHandle) {
             resetSecurityContext = SecurityContext.getCurrent();
             SecurityContext.setCurrent(invocationCtx.getSecurityContext());
         }
-        if (invocation != null) {
-            // Each invocation needs a ResourceTableKey that returns a unique hashCode for TransactionManager
-            invocation.setResourceTableKey(new PairKey(invocation.getInstance(), Thread.currentThread()));
-            invocationManager.preInvoke(invocation);
+
+        if (setup.isUnchanged(JNDI) || (!setup.isPropagated(JNDI) && !setup.isClear(JNDI))) {
+            ComponentInvocation currentInvocation = invocationManager.getCurrentInvocation();
+            if (currentInvocation != null) {
+                if (currentInvocation.getAppName() == null) {
+                    currentInvocation.setAppName(invocation.getAppName());
+                }
+                invocation = currentInvocation;
```

Comment:
It's safe to replace invocation with current invocation here? Or we should more carefully copy only some state from current invocation?

Reply:
The idea is that if application context is not cleared or propagated, it shouldn’t be changed if it’s already set. That’s according to the concurrency spec. That’s why I replace the new invocation with an existing one, if it’s present.

Reply:
Thank you for clarification.